### PR TITLE
Avoid redundant tombstones in secondary indexes

### PR DESCRIPTION
**Pull request by jamii**
I [noticed](https://github.com/tigerbeetledb/tigerbeetle/issues/262#issuecomment-1341623199) that a simple workload of transfers between two accounts was spending most of it's time compacting the balance trees. The reason is that updating the balance generates a slew of useless tombstones which are written to the tree and fill up level 0.

It's safe to drop these tombstones in the secondary indexes, but not in general. So let's add a parameter to `TableType`: `usage: enum {general, secondary_index}`.

Setting `usage == .secondary_index` promises:
* Only put a key if it doesn't exist.
* Only remove a key if it exists.
* Never look at `TableValue`, only `TableKey`. 

This is true for all of our index trees.

What we gain from this is that we know that for a given key the sequence of operations alternates puts and removes, and whenever we see a put and a remove together we can cancel them out.

This change increases throughput on the 2 account benchmark from `44523` tx/s to `159744` tx/s.

It was difficult to put the put/remove cancel into KWayMergeIterator, so I made a MergeIterator that is specialized for compaction. This cuts merge time a little (before/after below). We'd probably have to rework the TableIterator interface to do better (it forces a lot of repeated computation) but it's not nearly a priority yet.

image

### CODE REVIEW

**Comment by jamii** [comment]
I wanted to test secondary lookups in forest_fuzz too, but it looks like we don't actually have any code for secondary lookups in groove at the moment? Is it safe to go direct to the tree and skip prefetch?

**Comment by sentientwaffle** [comment]
:+1: Adding secondary index lookups will be part of the query API / range queries; it isn't implemented yet.

> Is it safe to go direct to the tree and skip prefetch?

Yes

**Comment by jamii** [comment]
Fuzzing with the forest is blocked by https://github.com/tigerbeetledb/tigerbeetle/issues/309, but I've tweaked the tree fuzzer to switch betwen .general and .secondary_index.

**Comment by jamii** [comment]
~@sentientwaffle This is ready for review now.~

Sorry about the gross merge. It was a nightmare to rebase when I'd changed the indentation of Environment.

**Comment by jamii** [comment]
Hmm, actually seeing some fuzzer failures :S

**Comment by jamii** [comment]
I'm struggling with this fuzzer crash. 

Here's the minimal change that triggers it:

```
git checkout origin/jamii-sanity
zig build fuzz_lsm_tree -- --seed 8902098390957168295
```

But it crashes on an assert inside k_way_merge, and nothing in that branch seems like it should be able to cause that. So I think I might have discovered an existing bug whose workload is only reachable after these changes.

**Comment by sentientwaffle** [comment]
I worked on it a bit but didn't narrow it down much. I agree that it doesn't appear to be directly related to your changes and is probably a preexisting bug that is only now being exposed.

This is what I've figured out so far:

- The keys it fails on are `1800825` and `1395874` — these are both from the same level (the "B" of the compaction) — I think it is level 0 but I didn't specifically check.
- `1395874` is the first key of a data block
- `1800825` is the last key of a data block
- For some reason our iterator is hitting `1800825` then `1395874`.
- This occurs after the "A" iterator has been emptied.

**Comment by jamii** [comment]
In that case, do you want to review this pr as is and I'll make a separate issue for the crash and tackle that next.

**Comment by sentientwaffle** [comment]
> In that case, do you want to review this pr as is and I'll make a separate issue for the crash and tackle that next.

:+1:

File: src/lsm/merge_iterator.zig
Comment (sentientwaffle) [inline src/lsm/merge_iterator.zig]: Is this the same as
```suggestion
                                assert(Table.tombstone(&value_a) != Table.tombstone(&value_b));
```

File: src/lsm/table_mutable.zig
Comment (sentientwaffle) [inline src/lsm/table_mutable.zig]: Would this refactor do the same thing?:
```zig
            if (upsert.found_existing) {
                if (usage == .secondary_index) {
                    // Put and remove cancel each other out.
                    assert(tombstone(upsert.key_ptr));
                    _ = table.values.remove(value.*);
                } else {
                    upsert.key_ptr.* = value.*;
                }
            }
```
Since if `!found_existing` then the `getOrPutAssumeCapacity` inserted the value, so we don't need to update `key_ptr`. (As a bonus this removes the compound condition, per TIGER_STYLE).

Ideally we can do likewise in `remove()`.

File: src/lsm/table_mutable.zig
Comment (sentientwaffle) [inline src/lsm/table_mutable.zig]: Updating the cache here contradicts `values_cache`'s field comment:
```zig
        /// The values cache is updated (in bulk) when the mutable table is sorted and frozen,
        /// rather than updating on every `put()`/`remove()`.
        /// This amortizes cache inserts for hot keys in the mutable table, and avoids redundantly
        /// storing duplicate values in both the mutable table and values cache.
```

Maybe only update the cache for `usage == .secondary_index` and add a note here explaining why we need to update the cache? (Also, I don't understand why the bulk-update doesn't work here. Are we exposing a stale value somehow?)

File: src/lsm/merge_iterator.zig
Comment (sentientwaffle) [inline src/lsm/merge_iterator.zig:44]: Document that returning `null` may indicate that _either_ the iterator is empty, or that it is drained.

File: src/lsm/merge_iterator.zig
Comment (jamii) [reply src/lsm/merge_iterator.zig]: :+1:

File: src/lsm/table_mutable.zig
Comment (jamii) [reply src/lsm/table_mutable.zig]: I ended up making it a lot more verbose, but it's obviously exhaustive now and the symmetry between put and remove is much clearer.

File: src/lsm/table_mutable.zig
Comment (jamii) [reply src/lsm/table_mutable.zig]: > I don't understand why the bulk-update doesn't work here

The existing code already updated the cache during put but not during remove. This was not a bug because get always checks values first before the cache. But if we remove the values entry instead of replacing it with a tombstone then we get a stale read from the cache.

If the cache is only supposed to be updated in bulk, we should probably remove the update in both put and remove?

Comment (sentientwaffle) [reply src/lsm/table_mutable.zig]: Ah, yeah. I don't think it matters too much, but I do lean a little towards bulk-update, since otherwise `put`s are overwriting values in the cache that could be useful later in the batch — but it doesn't actually get any use from those `put` values since they are also in the mutable table.

File: src/lsm/merge_iterator.zig
Comment (jamii) [reply src/lsm/merge_iterator.zig:44]: :+1: 

I actually don't like this design (too much branching, plus errors are more expensive than they look), but it matches k_way_merge and I'll refrain from messing with the iterators too much until merging becomes a bottleneck.

Comment (sentientwaffle) [reply src/lsm/merge_iterator.zig:44]: :+1: I'm definitely open to using a different approach. Prior to using errors to signal `Empty`/`Drained` we just required checking `is_empty()`/`is_drained()` predicates before calling `pop()`, but that was easy to forget & hard to debug.

File: src/lsm/table_mutable.zig
Comment (jamii) [reply src/lsm/table_mutable.zig]: Related - it doesn't seem that the cache knows about tombstones. Does it make sense to fill the cache with tombstones, or would it be better to treat those as a cache remove?

Comment (jamii) [reply src/lsm/table_mutable.zig]: I tentatively removed the cache updates in put and remove and so far it doesn't cause any fuzzing failures.

Comment (sentientwaffle) [reply src/lsm/table_mutable.zig]: :thinking: Removing probably makes the most sense. For our workload in particular removes from the object tree (which is what this cache covers) will almost always be followed by a `put`. Workloads in general might want negative-space caching, but the filter blocks + grid cache can cover that, so I'm not too worried.

File: src/lsm/merge_iterator.zig
Comment (jamii) [reply src/lsm/merge_iterator.zig:44]: I can imagine. Once it's blocking perf we'll probably want to make it either push-based, or keep it pull-based but return `*const [C]Value` instead of `Value` to amortize the ovehead of the iterator control flow.

**Review by sentientwaffle — APPROVED** [review]

**Pull request merged**
